### ISSUE TITLE
Add the source of the documentation

### ISSUE THREAD

**Issue by bourtemb** [tango-controls/TangoSourceDistribution#40]
Hello could you re-add the source of the documentation, in order to be able to re-generate the documention during the build process on the Debain bbuild farm.

thanks

_Originally posted by @picca in https://github.com/tango-controls/TangoSourceDistribution/pull/18#issuecomment-497922380_

**Comment by mliszcz** [comment]
Hi @picca, @bourtemb. Could you clarify what exactly is needed here?

* Do we need PDF sources (doc/ directory) in TangoSourceDistribution source tree?
* or do we need PDF sources in the actual package (tango-9.3.3-rc1.tar.gz)?
* Shall Ant build the PDF during source package generation?
* or shall the autotools build the PDF during package configuration/installation?

**Comment by picca** [comment]
Hello,
you can look at the tango 9 source distribution to understand. We need the source of the pdf and the autotools build the pdf from these files via lyx. It is available in the tango9 sources distribution.

please do not make ant mandatory to build the c++ part or the documentation.

**Comment by mliszcz** [comment]
I've found doc source in debian build repo (Imported Upstream version 9.2.5a+dfsg1): https://salsa.debian.org/science-team/tango/tree/06eac5ac99900e210cbddb6a258b66086b1f45c2/doc.

But I don't see these sources in upstream: https://github.com/tango-controls/TangoSourceDistribution/tree/Tango-Release-9.2.5a/doc.

I assume we'd need to take sources from https://github.com/tango-controls/tango-doc/. @picca can you confirm? If you need lyx files, I think it will be complicated, because currently the documentation is built on rtd from RST files, which are translated to latex with sphinx and then compiled to PDF. See: https://readthedocs.org/projects/tango-controls/builds/9182718/

@bourtemb, @Ingvord, @t-b, could you share your opinion if we really need to have doc sources inside the source distribution package? @picca maybe there should be a separate package for 'tango-doc'?

Tomorrow I'm OOO, I can work on this starting from friday if there is no other volunteer and if we agree on a way forward.

**Comment by picca** [comment]
Hello, In that case we juste need to adpat the automake rules in order to use sphinx in order to build the doc. :). Addind a new package just for the source is not a good idea most of the time :)

**Comment by mliszcz** [comment]
So, could you confirm that what you need is to be able to do something like:
1. install sphinx and latex
1. download and extract tango-9.3.3-rc1.tar.gz
1. run `./configure && make pdf && make install-pdf`

Alternatively, the last step could be: `./configure --with-pdf && make && make install` - which one is better?

**Comment by picca** [comment]
Exactly, but maybe we can switch to the html doc and not the pdf one.
so in that case  make we can just put the doc in the sourcedirectory  and use the sphinx-build script in order to generate the documentation.

In order to facilitate the bootstrap of new architectures, it is important to be able to  build a package with less dependencies. So having a way to not build the doc would be nice.
In that case maybe the best solution is to use sphinx-build in order to build the doc and not to integrate it into the autotools.

**Comment by t-b** [comment]
@picca @mliszcz I'm not a debian package maintainer but a user for a long time.

In debian the sources of the documentation must be available to be able to regenerate the output formats (PDF in this case). So I would say whichever package (tango-9.3.3-rc1.tar.gz or the source distribution) ships the PDF documentation should also ship the source and necessart scripts to regenerate the PDF.

**Comment by mliszcz** [comment]
Thanks @picca, @t-b.

> Exactly, but maybe we can switch to the html doc and not the pdf one.

Indeed it is easier to build the html documentation. You need only Sphinx and IPython.

However I'm not in a position to remove the pdf file from TangoSourceDistribution or to decide whether we shall go with pdf or html.

> So having a way to not build the doc would be nice.
In that case maybe the best solution is to use sphinx-build in order to build the doc and not to integrate it into the autotools.

I've opened PR #42 with doc sources added back to the tarball. It is not integrated with autotools (as you proposed, also this is in-line with our plans to remove autotools in future).

There is a requirements.txt file that lists Sphinx and IPython. There is also a makefile for convenience. When preparing the .deb package, you need to remove the pdf and copy html docs to the installation directory.

```bash
./configure --prefix=/prefix --disable-dbserver --disable-dbcreate --disable-java
make -j4
sudo make install

# ------ extra step for debian packaging
pip install --user -r doc/src/requirements.txt
make -C doc/src/ singlehtml

sudo rm /prefix/share/doc/tango/*.pdf
sudo cp -r doc/src/build/singlehtml/* /prefix/share/doc/tango/
# ------ extra step end

firefox /prefix/share/doc/tango/contents.html
```

@picca, is this ok for you?

**Comment by bourtemb** [comment]
> > Exactly, but maybe we can switch to the html doc and not the pdf one.
> 
> Indeed it is easier to build the html documentation. You need only Sphinx and IPython.
> 
> However I'm not in a position to remove the pdf file from TangoSourceDistribution or to decide whether we shall go with pdf or html.

> I've opened PR #42 with doc sources added back to the tarball. It is not integrated with autotools (as you proposed, also this is in-line with our plans to remove autotools in future).

> There is a requirements.txt file that lists Sphinx and IPython. There is also a makefile for convenience. When preparing the .deb package, you need to remove the pdf and copy html docs to the installation directory.

I think we should simply remove the pdf from the source distribution and provide the html documentation with the rst files instead.

**Comment by bourtemb** [comment]
Sorry, the previous comment was posted by mistake.
I edited it.

**Comment by mliszcz** [comment]
Thanks @bourtemb. Let me summarize the changes:
* `ant build package`
  * will no longer pull pdf file into `doc/`
  * will pull tango-doc source tree into `doc/src/`
  * will build 'singlehtml' documentation in `doc/html/`
  * **this is extra dependency on python and sphinx to create the tarball**
  * impacts: `build.xml`
* `configure; make; make install`
  * will copy the docs like: `cp -r doc/html/* $PREFIX/share/doc/tango/`
  * impacts: `assets/doc/Makefile.am`

If it's ok, I can apply these changes to the PR.

**Comment by bourtemb** [comment]
> Let me summarize the changes:
> 
>     * `ant build package`
>       
>       * will no longer pull pdf file into `doc/`
>       * will pull tango-doc source tree into `doc/src/`
>       * will build 'singlehtml' documentation in `doc/html/`
>       * **this is extra dependency on python and sphinx to create the tarball**
>       * impacts: `build.xml`
> 
>     * `configure; make; make install`
>       
>       * will copy the docs like: `cp -r doc/html/* $PREFIX/share/doc/tango/`
>       * impacts: `assets/doc/Makefile.am`
> 
> 
> If it's ok, I can apply these changes to the PR.

Good summary!
Looks good to me!

**Comment by picca** [comment]
Doest it means, that I need ant in order to build the documentation ?.

**Comment by bourtemb** [comment]
@picca, I think you will use the tarball directly in your case.
ant is necessary to build the tarball.
But we will provide the tarball with the documentation already generated (as well as the rst files needed to eventually rebuild the documentation).

**Comment by picca** [comment]
Yes, I need to rebuild the doc, so it would be nice to hqve q target in the makefile.am in order to build the documentation.

**Comment by mliszcz** [comment]
@bourtemb, changes are applied in #42. I have rebased the PR on `master`

@picca, you proposed the opposite in one of your previous posts. Can you use `sphinx-build` (or makefile from `doc/src/`) directly to build the documentation?
> In that case maybe the best solution is to use sphinx-build in order to build the doc and not to integrate it into the autotools.

**Comment by picca** [comment]
Sorry If I confused, you, I meant that it would be nice to have the same recipe in order to build the documentation (upstream and maintainer). 

It seems to me that there is two way to build the documentation, one from ant and another one once you have the sources and which use sphinx-build.

If you do not want to be bothered, let's first build the doc from sphinx-build with no integration in the autotools, and we will see later fot eh next step.

Cheers

**Comment by mliszcz** [comment]
@picca, let me clarify a few things:
* html documentation is included in source tarball, so that the user of the tarball does not need python or sphinx to compile tango from source,
* **there is just one way to build the documentation** and it is `sphinx-build`. `sphinx-build` is run by `ant` when the source tarball is being built,
* in your case, the 'upstream' is probably the source tarball, which does not need ant to build anything,
* 'upstream' provides pre-built documentation and sources so that you can re-build it if you want.

I'm not sure if we want to integrate the doc build process with autotools. The doc source is provided only to satisfy debian packaging rules. Most of the users will be fine with pre-compiled documentation. Any additional step during tango installation is a potential point of failure (i.e. user has no Sphinx installed, the version is incorrect, etc. ...).

**Comment by picca** [comment]
I get the points. So for now just tell me how to build the doc and I will tune the Debian packaging in order to re-generate the documentation. Debian makes no difference between documentation sources and code sources. We rebuild everything and provide this guaranty to our users. I think that this is almost the same with the other Distributions.

**Comment by mliszcz** [comment]
If #42 is merged in it's current state, below steps should work for you:
```bash
./configure --prefix=/prefix --disable-dbserver --disable-dbcreate --disable-java
make -j4
sudo make install

sudo rm -rf /prefix/share/doc/tango/html # remove pre-compiled doc
pip install --user Sphinx ipython
make -C ./doc/src/ singlehtml
sudo cp -r ./doc/src/build/singlehtml /prefix/share/doc/tango/html

**Comment by picca** [comment]
Ok, for me

**Comment by mliszcz** [comment]
@picca, just discovered one more thing. The *source* of the documentation has some files which *may* be considered binary/compiled/generated:

* oher PDFs:
https://github.com/tango-controls/tango-doc/blob/master/source/overview/introduction/tango_basics.pdf
* presentations:
https://github.com/tango-controls/tango-doc/blob/master/source/tools-and-extensions/archiving/HDB-plus-plus/hdbcassandra-tango-workshop.pptx
* text documents
https://github.com/tango-controls/tango-doc/blob/master/source/tools-and-extensions/archiving/src-materials/Bensikin%20Manual_v1.1.doc
* image projects
https://github.com/tango-controls/tango-doc/blob/master/source/development/advanced/rest-api/ESRF.odg
* source maps
https://github.com/tango-controls/tango-doc/blob/master/source/_theme/sphinx_rtd_theme/static/css/theme.css.map

Will these be a problem for you?

**Comment by picca** [comment]
tango_basics.pdf -> sources ?
hdbcassandra-tango-workshop.pptx -> removed ?
Bensikin%20Manual_v1.1.doc -> removed ?
ESRF.odg -> this is the prefered source format for modification ?
theme.css.map -> this file is part of the sphinx_rtd_theme, maybe this could be remove from the doc.

I suppose that you the tango-doc repository can be see has the source for the documentation so in that case having sphinx is not an issue. Ant it is best to remove all the generated files.

**Comment by mliszcz** [comment]
> tango_basics.pdf -> sources ?

There is a bunch of pdf files we don't have the sources for, e.g:

```
./overview/introduction/tango_basics.pdf
./overview/introduction/tangooverview.pdf
./overview/introduction/tango_introduction.pdf
./administration/services/hdbpp/hdb-design-guidelines-18.pdf
./tools-and-extensions/archiving/HDB-plus-plus/20160510-hdb-meeting.pdf
./tools-and-extensions/archiving/HDB-plus-plus/proceedings_ICALEPCS2015_cassandra.pdf
./tools-and-extensions/archiving/HDB-plus-plus/proceedings_ICALEPCS2015.pdf
./tools-and-extensions/built-in/astor/astor.pdf
./tools-and-extensions/built-in/pogo/THP080.pdf
./tools-and-extensions/gui/qtango/qtangopresentation.pdf
./tools-and-extensions/gui/qtango/qtwatcherqtwriter.pdf
./tools-and-extensions/gui/qtango/thp096.pdf
./tutorials-and-howtos/tutorials/tango_training.pdf
./tutorials-and-howtos/tutorials/doctpv2.pdf
./tutorials-and-howtos/tutorials/ai_5_scada_light.pdf
./tutorials-and-howtos/tutorials/tango_training_python.pdf
./tutorials-and-howtos/tutorials/tango_class_arduino_temp_sensor.pdf
```

> ESRF.odg -> this is the prefered source format for modification ?

I don't know.

> hdbcassandra-tango-workshop.pptx -> removed ?
> Bensikin%20Manual_v1.1.doc -> removed ?

I'm not sure if we should remove anything. How can we decide what is important and what can be removed without any harm / impact on docs quality?

> theme.css.map -> this file is part of the sphinx_rtd_theme, maybe this could be remove from the doc.

`sphinx_rtd_theme` is needed for html documentation to work, right? If removed, we will lost e.g. some CSS.

@bourtemb what is your opinion? I'd like to avoid tampering with tango-doc source in TangoSourceDistribution more than necessary. Maybe the issue shall go to tango-doc to make the soures debian-friendly?

**Comment by mliszcz** [comment]
Hi @picca and @bourtemb 

I'd like to close #42 finally and need your decision how we'd like to proceed.

I've added below action after tango-doc checkout during ant build:

```diff
+        <delete>
+            <fileset dir="${workdir}/tango-doc/source">
+                <include name="**/*.pdf"/>
+                <include name="**/*.doc"/>
+                <include name="**/*.docx"/>
+                <include name="**/*.pptx"/>
+                <include name="**/*.odg"/>
+                <include name="**/*.class"/>
+            </fileset>
+        </delete>
```

It works but throws some warnings during doc compilation, e.g.:
```
/home/tango-cs/build/tango-9.3.3/doc/src/source/tools-and-extensions/archiving/HDB++.rst:314: WARNING: download file not readable: /home/tango-cs/build/tango-9.3.3/doc/src/source/tools-and-extensions/archiving/HDB-plus-plus/20160510-hdb-meeting.pdf
```

Of course links to excluded resources are broken in generated documentation.

Should I apply this change to the PR?

**Comment by bourtemb** [comment]
It is clear that we cannot include pdf files within the Debian package without the sources to generate it.
So one solution is indeed to remove them from the package since we don't have anymore the source files to generate some of these pdf files...
Another solution is to provide a link to a package already online instead. I guess this should not be forbidden by the Debian policies?

It looks like some .doc files are actually text files (doc/src/source/development/device-api/device-server-writing/java_exchanging/ty.doc).

I don't understand why we have a .class file (doc/src/source/development/device-api/device-server-writing/java_exchanging/ty.class) in the doc...
At first sight, it looks like the files in ./development/device-api/device-server-writing/java_exchanging directory are not used.  @piogor , do you know whether these files are used?
It's also interesting to note that .doc files in this _java_exchanging_ directory are not binary files.

@picca, can you confirm that we cannot include .doc, .docx, .pptx, .odg files? There is no source needed to generate these files...
I understand why we need the source of the pdf and ps files, but the .doc, .docx, .pptx and .odg files, they are not generated from a text file... (doc and docx files could even be considered as text files...).
But perhaps there is a license issue with these files?
Any clarification would be appreciated.

**Comment by picca** [comment]
Hello, all libreoffice format are allowed if they are the prefered format for modification.
I found at least one package in debian main which contain a docx (lib3mf) so it seems that it works for the [doc|ppt]x familly.
For the doc format, I would say for now no, but I maybe wrong.

Maybe the best would be to migrate to libreoffice format to simplify the maintenance.
Or better to switch all the documentation to .rst as a long time project :)

Cheers

**Comment by mliszcz** [comment]
Hi all,
I've updated #42 to remove all pdf, doc and class files. docx, pptx and odg are left.

@bourtemb the files from `java_exchanging` directory are used in old Tango documentation hosted at ESRF webpage:
http://www.esrf.eu/computing/cs/tango/tango_doc/kernel_doc/ds_prog/node9.html
```
any.doc -> 8.2.2.3.2 Inserting/Extracting TANGO sequences, structures or enumeration
any_ins.doc -> 8.2.2.4 The insert and extract methods of the Command class for Java
excep.doc -> 8.2.4.2 Example of throwing exception using Java
ty.doc -> 8.2.1.2.4 Enumeration
```
Probably they were never referenced in tango-doc repository

**Comment by bourtemb** [comment]
> Hi all,
> I've updated #42 to remove all pdf, doc and class files. docx, pptx and odg are left.

@picca, I was wondering, wouldn't it make more sense to have a separate Debian package tango-doc for the documentation and to add it as recommended/suggested package and to simply add a link to the online doc on readthedoc in the README of tango packages?
I think not everybody will be interested in having the doc locally and we've seen the doc can be relatively big since there are several files like presentations included in it and I guess the size might grow in the future.

> 
> @bourtemb the files from `java_exchanging` directory are used in old Tango documentation hosted at ESRF webpage:
> http://www.esrf.eu/computing/cs/tango/tango_doc/kernel_doc/ds_prog/node9.html
> 
> ```
> any.doc -> 8.2.2.3.2 Inserting/Extracting TANGO sequences, structures or enumeration
> any_ins.doc -> 8.2.2.4 The insert and extract methods of the Command class for Java
> excep.doc -> 8.2.4.2 Example of throwing exception using Java
> ty.doc -> 8.2.1.2.4 Enumeration
> ```
> 
> Probably they were never referenced in tango-doc repository

Good catch.
Actually, it seems that these sections got unreferenced when passing from Tango 8 to Tango 9.

**Comment by picca** [comment]
To my opinion, it is better to provide the documentation with the software. The doc is part of the software. The release manager should reject PR if the documentation is not updated with the rest of the code. This way the doc is always up to date with the state of the software.

but all this is only my view.

It the tango project consider that documentation should be in it his own repository it is up to you.
Then we need to create another source package for the documentation.
It is worst because we need to update two source package instead of one when upgrading tango.
so this is more work. (it is most of the time bad to require more work from others ;)

Cheers

**Comment by mliszcz** [comment]
@picca, side question - do you plan to release just one debian package (based on source distribution) with everything included - library, docs, device servers, java apps, etc.?

**Comment by picca** [comment]
In fact there is two kind of packaes in Debian
* source packages
* binary packages.

I have one source pacakge for tango, whcih generate many binary packages.

look at here
https://tracker.debian.org/pkg/tango

I can not distribut the java part because of an issue with jacorb, but all the rest is ok.
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=349540

Cheers

**Comment by mliszcz** [comment]
Thanks! I was just concerned that there will be one binary package starting from 9.3, but I was wrong.

**Comment by bourtemb** [comment]
> I have one source pacakge for tango, whcih generate many binary packages.
> 
> look at here
> https://tracker.debian.org/pkg/tango

Thanks Frédéric for the explanations and this very useful link.
So there is already a separate libtango-doc package (which is generated from the source package).
Up to now, this package was containing mainly the tango book as pdf (no presentation).
There has been quite some changes since then because the tango-doc github repository has been created and the documentation has been migrated to readthedocs.
During this process, the Tango book has been integrated into readthedocs as well as many other sources of information related to Tango (including presentations) so the doc is much bigger than before.
The presentations were not provided in the libtango-doc package.
So if we remove the files as done by @mliszcz in #42 we are still providing all the documentation which was in the Tango book and even a bit more. It's not very clean because some links will fail in the documentation but we provide even more information than we used to do in the previous versions of libtango-doc. So this might be the way to go for the moment, until we convert the problematic files to a Debian compliant format or we find the source of the pdf files.

> ```
> any.doc -> 8.2.2.3.2 Inserting/Extracting TANGO sequences, structures or enumeration
> any_ins.doc -> 8.2.2.4 The insert and extract methods of the Command class for Java
> excep.doc -> 8.2.4.2 Example of throwing exception using Java
> ty.doc -> 8.2.1.2.4 Enumeration
> ```
> 
> 
> Probably they were never referenced in tango-doc repository

It seems like these files got unreferenced from Tango 8 to Tango 9 because the Java servers implementation was reimplemented and these documentation files became obsolete.
I suggest to remove them from tango-doc github repo.

**Comment by mliszcz** [comment]
FYI, I've deleted `java_exchange` dir in tango-controls/tango-doc#282

**Comment by mliszcz** [comment]
Hi, @bourtemb @picca, do you have any updates on 9.3.3 final release date (including debian package)? - I assume it is blocked by #40 and #41, for which PRs are ready since a few weeks.

**Comment by bourtemb** [comment]
Sorry @mliszcz .
I have no date. I need to find some time to review and merge #40 and #41.  I will do that this week.
We need a new release of JTango too because the version we are providing in 9.3.3-rc1 contains an annoying bug (possible deadlock) which seems to have been fixed in jtango-9-lts branch.

**Comment by mliszcz** [comment]
Thanks @bourtemb!
I was just asking because we're waiting to tag as 9.3.3 things like docs, vm, etc.

CC @piogor

**Comment by bourtemb** [comment]
@mliszcz 
I have created https://github.com/mliszcz/TangoSourceDistribution/pull/1 to update assets/README too which was still referring to tango.pdf file in the DOCUMENTATION section.

**Comment by mliszcz** [comment]
@bourtemb thanks!

**Comment by bourtemb** [comment]
Fixed by #42
